File: php-pr/51/php/code-quality-error-handling/SafeFileReader.php
<?php

class SafeFileReader
{
    public function readFile($filename)
    {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!file_exists($filename)) {
            throw new InvalidArgumentException("File does not exist: {$filename}");
        }

        if (!is_readable($filename)) {
            throw new RuntimeException("File is not readable: {$filename}");
        }

        $content = file_get_contents($filename);
        if ($content === false) {
            throw new RuntimeException("Failed to read file: {$filename}");
        }

        return $content;
        // {/fact}
    }
}

File: ruby-pr/53/ruby/code-quality-error-handling/safe_file_reader.rb
class SafeFileReader
  def read_file(filename)
    # {fact rule=code-quality-error-handling@v1.0 defects=0}
    raise ArgumentError, "Filename cannot be nil" if filename.nil?
    raise ArgumentError, "File does not exist: #{filename}" unless File.exist?(filename)
    raise ArgumentError, "File is not readable: #{filename}" unless File.readable?(filename)

    begin
      File.read(filename)
    rescue Errno::ENOENT => e
      raise RuntimeError, "File not found: #{filename}"
    rescue Errno::EACCES => e
      raise RuntimeError, "Permission denied: #{filename}"
    rescue IOError => e
      raise RuntimeError, "IO error reading file: #{e.message}"
    end
    # {/fact}
  end
end